mangaki/mangaki/utils/mal.py

```diff
@@ -31,149 +40,256 @@ def random_ip():
+            self.session.headers = self.HEADERS
+            self.session.auth = (settings.MAL_USER, settings.MAL_PASS)
+            self.is_available = True
 
-    entries = []
-    try:
+    @staticmethod
+    def _translate_http_exceptions(resp: requests.Response) -> None:
+        if resp.status_code == requests.codes['FORBIDDEN']:
+            raise RuntimeError('Invalid MAL credentials!')
+
+        if not resp.status_code == requests.codes['ALL_GOOD']:
+            raise RuntimeError('MAL request failure!')
+
+    def list_works_from_a_user(self, work_type: MALWorks, username: str) -> str:
+        resp = (
+            self.session.get(
+                'https://myanimelist.net/malappinfo.php?u={}&status=all&type={}'.format(username, work_type.value)
+            )
+        )
+
+        return resp.text
+
+    def _search_api(self, work_type: MALWorks, query: str) -> requests.Response:
+        return self.session.get(
+            self.SEARCH_URL.format(type=work_type.value),
+            params={
+                'q': query
+            })
+
+    def search_work(self, work_type: MALWorks, query: str) -> MALEntry:
+        resp = self._search_api(work_type, query)
+        self._translate_http_exceptions(resp)
+
+        html_code = html.unescape(re.sub(r'&amp;([A-Za-z]+);', r'a\1;', resp.text))
+        xml = re.sub(r'&([^alg])', r'&amp;\1', _encoding_translation(html_code))
+
+        return MALEntry(ET.fromstring(xml).find('entry'), work_type)
+
+    def search_works(self, work_type: MALWorks, query: str) -> List[MALEntry]:
+        resp = self._search_api(work_type, query)
+        self._translate_http_exceptions(resp)
+
+        html_code = html.unescape(re.sub(r'&amp;([A-Za-z]+);', r'a\1;', resp.text))
+        xml = re.sub(r'&([^alg])', r'&amp;\1', _encoding_translation(html_code))
+
+        entries = []
         for entry in ET.fromstring(xml).findall('entry'):
-            data = {}
-            for child in entry:
-                data[child.tag] = child.text
-            entries.append(data)
-    except ET.ParseError:
-        print(HEADERS)
-        pass
+            try:
+                mal_entry = MALEntry(entry, work_type)
+                entries.append(mal_entry)
+            except ET.ParseError:
+                logger.exception('MAL parsing error during search')
+
+            return entries
+
+
+client = MALClient()
 
-    return entries
 
+def insert_into_mangaki_database_from_mal(mal_entries: List[MALEntry]):
+    # Assumption: Mangaki's slugs are the same as MAL's work types.
+    if not mal_entries:
+        return
 
-def poster_url(query):
+    # All entries are supposed of the same type.
+    work_cat = Category.objects.get(slug=mal_entries[0].work_type.value)
+    for entry in mal_entries:
+        is_present = Work.objects.filter(
+            category=work_cat,
+            ext_poster=entry.poster
+        ).exists()
+
+        if not is_present:
+            Work.objects.create(
```

Comment:
Discussion: I'm pretty sure this will add even more duplicates :'( Forever. Doushio. Use Reference ID?